apps/fipsinstall.c

```diff
@@ -0,0 +1,350 @@
+#include "apps.h"
+#include "progs.h"
+#include <openssl/evp.h>
+#include <openssl/provider.h>
+
+#define BUFSIZE 4096
+#define DEFAULT_MAC_NAME "HMAC"
+
+/* Configuration file values */
+#define FIPS_SECTION "[fips_check_section]"
+#define VERSION_KEY  "version"
+#define VERSION_VAL  "1"
+#define MODULE_MAC_KEY "module.mac"
+#define INSTALL_MAC_KEY "install.mac"
+#define INSTALL_STATUS_KEY "install.status"
+#define INSTALL_STATUS_VAL "INSTALL_SELF_TEST_KATS_RUN"
+
+typedef enum OPTION_choice {
+    OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+    OPT_IN, OPT_OUT, OPT_MACOPT, OPT_VERIFY,
+    OPT_MAC
+} OPTION_CHOICE;
+
+const OPTIONS fipsinstall_options[] = {
+    {"help", OPT_HELP, '-', "Display this summary"},
+    { OPT_MORE_STR, 0, 0, "e.g: openssl fipsinstall -in libfips.so -out "
+      "fips.conf -mac HMAC -macopt digest:SHA256 -macopt hexkey:00"},
+    {"in", OPT_IN, 's', "Input file"},
+    {"out", OPT_OUT, '>', "Output file"},
+    {"mac", OPT_MAC, 's', "MAC name"},
+    {"macopt", OPT_MACOPT, 's', "MAC algorithm control parameters in n:v form. "
+                                "See 'Supported Controls' in the EVP_MAC_ docs"},
+    {"verify", OPT_VERIFY, '-', "Verify the installed status is correct"},
+
+    {NULL}
+};
+
+static int mac_ctrl_string(EVP_MAC_CTX *ctx, const char *value)
+{
+    int rv;
+    char *stmp, *vtmp = NULL;
+
+    stmp = OPENSSL_strdup(value);
+    if (stmp == NULL)
+        return -1;
+    vtmp = strchr(stmp, ':');
+    if (vtmp != NULL) {
+        *vtmp = 0;
+        vtmp++;
+    }
+    rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+    OPENSSL_free(stmp);
+    return rv;
+}
+
+static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
+                  unsigned char *out, size_t *out_len)
+{
+    int ret = 0;
+    int i;
+
+    if (!EVP_MAC_init(ctx))
+        goto err;
+    if (EVP_MAC_size(ctx) > *out_len)
+        goto end;
+    for (;;) {
+        i = BIO_read(in, (char *)tmp, BUFSIZE);
+        if (i < 0)
+            goto err;
+        if (i == 0)
+            break;
+        if (!EVP_MAC_update(ctx, tmp, i))
+            goto err;
+    }
+end:
+    if (!EVP_MAC_final(ctx, out, out_len))
+        goto err;
+    ret = 1;
+err:
+    return ret;
+}
+
+static int print_mac(BIO *bio, const char *label, unsigned char *mac,
+                     size_t len)
+{
+    size_t i;
+
+    if (!BIO_printf(bio, "%s = ", label))
+        return 0;
+    for (i = 0; i < len; ++i) {
+        if (!BIO_printf(bio, "%02X", mac[i]))
+            return 0;
+    }
+    return BIO_printf(bio, "\n");
+}
+
+static int load_fips_prov_and_run_self_test(const char *module_filename)
+{
+    int ret = 0;
+    OSSL_PROVIDER *prov = NULL;
+
+    prov = OSSL_PROVIDER_load(NULL, module_filename);
+    if (prov == NULL) {
+        BIO_printf(bio_err, "Failed to load fips module\n");
+        goto end;
+    }
+    /*
+     * TODO - make sure self tests arte activated here
+     * It must return NULL if the self tests fails.
+     */
```

Comment:
It will be interesting to see what you come up with here.  Me, I'm imagining passing some sort of OSSL_PARAM that acts like a trigger in the receiving end, i.e. isn't really a value to be stored, but rather a control.  Was that what you were thinking of doing?

Reply:
Hmm the only thing I think I need to pass is the module checksum..
The other values are all loaded from config files. The trigger is that there is no fips install file yet (which contains a few extra values).. Do you mind a OSSL_PARAM * being added to the OSSL_PROVIDER_load (or a different API for this case??)

Reply:
I don't mind.  There's an internal routine `ossl_provider_add_parameter` that could help.  However, that one takes string values only, so you might want to think about how you want to handle that, given that an OSSL_PARAM array given to `OSSL_PROVIDER_load` can potentially contain anything.